File: LeetCode/medium/Remove_Nth_Node_From_End_of_List.cxx
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode *a = head, *b = head;
        ListNode *a_prev = nullptr;
        for (int i = 0; i < n-1; ++i)
        {
            b = b->next;
            if (!b)
                return head;
        }
        while (b->next)
        {
            a_prev = a;
            a = a->next;
            b = b->next;
        }
        
        if (a_prev == nullptr)
            return a->next;
        else
        {
            a_prev->next = a->next;
            return head;
        }
        
    }
};

File: LeetCode/medium/Remove_Nth_Node_From_End_of_List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        r = head

        l = r
        prev_l = None
        for i in range(n):
            r = r.next

        while r is not None:
            prev_l = l
            r = r.next
            l = l.next
        if prev_l is not None:
            prev_l.next = l.next
            return head
        else:
            return l.next

